File: electron/services/skill-parser.ts
import fs from "fs";
import path from "path";
import { logger } from "../utils/logger";

/**
 * Parsed skill definition from a SKILL.md file.
 */
export interface ParsedSkillDefinition {
  name: string;
  description: string;
  model?: string;
  tools?: string[];
  maxBudgetUsd?: number;
  instructions: string;
  rawContent: string;
}

/**
 * Parse a SKILL.md file content into a structured skill definition.
 *
 * Format:
 *   ---
 *   name: Skill Name
 *   description: >
 *     Multi-line description
 *   model: sonnet
 *   tools:
 *     - Read
 *     - Grep
 *   max_budget_usd: 0.50
 *   ---
 *
 *   ## Instructions
 *   ...body content...
 */
export function parseSkillMd(content: string): ParsedSkillDefinition {
  const trimmed = content.trim();

  // Check for frontmatter delimiters
  if (!trimmed.startsWith("---")) {
    throw new Error("SKILL.md must start with --- frontmatter delimiter");
  }

  const secondDelimiter = trimmed.indexOf("---", 3);
  if (secondDelimiter === -1) {
    throw new Error("SKILL.md missing closing --- frontmatter delimiter");
  }

  const frontmatter = trimmed.slice(3, secondDelimiter).trim();
  const body = trimmed.slice(secondDelimiter + 3).trim();

  // Parse YAML-like frontmatter (simple key-value parser)
  const parsed = parseSimpleYaml(frontmatter);

  const name = parsed.name;
  if (!name) {
    throw new Error("SKILL.md frontmatter must include 'name'");
  }

  const description = parsed.description ?? "";
  const model = parsed.model;
  const tools = parsed.tools as string[] | undefined;
  const maxBudgetUsd = parsed.max_budget_usd
    ? parseFloat(String(parsed.max_budget_usd))
    : undefined;

  return {
    name: String(name),
    description: String(description),
    model: model ? String(model) : undefined,
    tools: tools
      ? (Array.isArray(tools) ? tools.map(String) : [String(tools)])
      : undefined,
    maxBudgetUsd:
      maxBudgetUsd !== undefined && !isNaN(maxBudgetUsd)
        ? maxBudgetUsd
        : undefined,
    instructions: body,
    rawContent: content,
  };
}

/**
 * Simple YAML-like parser for frontmatter.
 * Handles: key: value, key: > multi-line, key: [list with - items]
 */
function parseSimpleYaml(
  yaml: string
): Record<string, string | string[] | undefined> {
  const result: Record<string, string | string[] | undefined> = {};
  const lines = yaml.split("\n");
  let currentKey = "";
  let currentValue = "";
  let inMultiline = false;
  let inList = false;
  let listItems: string[] = [];

  for (const line of lines) {
    const trimmedLine = line.trim();

    if (inList) {
      if (trimmedLine.startsWith("- ")) {
        listItems.push(trimmedLine.slice(2).trim());
        continue;
      } else {
        // End of list
        result[currentKey] = listItems;
        inList = false;
        listItems = [];
      }
    }

    if (inMultiline) {
      if (
        trimmedLine === "" ||
        (!line.startsWith(" ") && !line.startsWith("\t") && trimmedLine.includes(":"))
      ) {
        // End of multiline
        result[currentKey] = currentValue.trim();
        inMultiline = false;
      } else {
        currentValue += " " + trimmedLine;
        continue;
      }
    }

    // Parse key: value
    const colonIndex = trimmedLine.indexOf(":");
    if (colonIndex > 0) {
      currentKey = trimmedLine.slice(0, colonIndex).trim();
      const value = trimmedLine.slice(colonIndex + 1).trim();

      if (value === ">" || value === "|") {
        // Multiline value
        inMultiline = true;
        currentValue = "";
      } else if (value === "") {
        // Could be start of a list
        inList = true;
        listItems = [];
      } else {
        result[currentKey] = value;
      }
    }
  }

  // Handle trailing values
  if (inMultiline && currentKey) {
    result[currentKey] = currentValue.trim();
  }
  if (inList && currentKey) {
    result[currentKey] = listItems;
  }

  return result;
}

/**
 * Load and parse a SKILL.md file from disk.
 */
export function loadSkillFile(filePath: string): ParsedSkillDefinition | null {
  try {
    if (!fs.existsSync(filePath)) {
      return null;
    }
    const content = fs.readFileSync(filePath, "utf-8");
    return parseSkillMd(content);
  } catch (err) {
    logger.warn("skill-parser", `Failed to parse ${filePath}`, err);
    return null;
  }
}

/**
 * Write a skill definition to a SKILL.md file.
 */
export function writeSkillFile(
  dirPath: string,
  definition: ParsedSkillDefinition
): void {
  if (!fs.existsSync(dirPath)) {
    fs.mkdirSync(dirPath, { recursive: true });
  }

  let frontmatter = `---\nname: ${definition.name}\n`;
  frontmatter += `description: >\n  ${definition.description}\n`;
  if (definition.model) {
    frontmatter += `model: ${definition.model}\n`;
  }
  if (definition.tools && definition.tools.length > 0) {
    frontmatter += `tools:\n`;
    for (const tool of definition.tools) {
      frontmatter += `  - ${tool}\n`;
    }
  }
  if (definition.maxBudgetUsd !== undefined) {
    frontmatter += `max_budget_usd: ${definition.maxBudgetUsd}\n`;
  }
  frontmatter += `---\n\n`;

  const content = frontmatter + definition.instructions;
  const filePath = path.join(dirPath, "SKILL.md");
  fs.writeFileSync(filePath, content, "utf-8");
}

/**
 * Scan a directory for SKILL.md files (one level deep).
 * Returns array of { dirName, parsed } for each found skill.
 */
export function scanSkillsDirectory(
  basePath: string
): Array<{ dirName: string; path: string; parsed: ParsedSkillDefinition }> {
  const results: Array<{
    dirName: string;
    path: string;
    parsed: ParsedSkillDefinition;
  }> = [];

  if (!fs.existsSync(basePath)) {
    return results;
  }

  const entries = fs.readdirSync(basePath, { withFileTypes: true });
  for (const entry of entries) {
    if (!entry.isDirectory()) continue;

    const skillFile = path.join(basePath, entry.name, "SKILL.md");
    const parsed = loadSkillFile(skillFile);
    if (parsed) {
      results.push({
        dirName: entry.name,
        path: path.join(basePath, entry.name),
        parsed,
      });
    }
  }

  return results;
}

File: src-tauri/src/services/skill_parser.rs
use std::path::Path;

use serde::{Deserialize, Serialize};
use tracing::warn;

use crate::error::{AppError, AppResult};

// ── Types ──────────────────────────────────────────────────

/// Parsed skill definition from a SKILL.md file.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ParsedSkillDefinition {
    pub name: String,
    pub description: String,
    pub model: Option<String>,
    pub tools: Option<Vec<String>>,
    pub max_budget_usd: Option<f64>,
    pub instructions: String,
    pub raw_content: String,
}

/// Entry found when scanning a skills directory.
#[derive(Debug, Clone)]
pub struct ScannedSkill {
    pub dir_name: String,
    pub path: String,
    pub parsed: ParsedSkillDefinition,
}

// ── Public API ─────────────────────────────────────────────

/// Parse a SKILL.md file content into a structured skill definition.
///
/// Expected format:
/// ```text
/// ---
/// name: Skill Name
/// description: >
///     Multi-line description
/// model: sonnet
/// tools:
///     - Read
///     - Grep
/// max_budget_usd: 0.50
/// ---
///
/// ## Instructions
/// ...body content...
/// ```
pub fn parse_skill_md(content: &str) -> AppResult<ParsedSkillDefinition> {
    let trimmed = content.trim();

    if !trimmed.starts_with("---") {
        return Err(AppError::InvalidInput(
            "SKILL.md must start with --- frontmatter delimiter".into(),
        ));
    }

    let second_delimiter = trimmed[3..].find("---");
    let second_delimiter = match second_delimiter {
        Some(pos) => pos + 3,
        None => {
            return Err(AppError::InvalidInput(
                "SKILL.md missing closing --- frontmatter delimiter".into(),
            ));
        }
    };

    let frontmatter = trimmed[3..second_delimiter].trim();
    let body = trimmed[second_delimiter + 3..].trim();

    let parsed = parse_simple_yaml(frontmatter);

    let name = parsed
        .get("name")
        .and_then(|v| match v {
            YamlValue::Scalar(s) => Some(s.clone()),
            _ => None,
        })
        .ok_or_else(|| {
            AppError::InvalidInput("SKILL.md frontmatter must include 'name'".into())
        })?;

    let description = parsed
        .get("description")
        .and_then(|v| match v {
            YamlValue::Scalar(s) => Some(s.clone()),
            _ => None,
        })
        .unwrap_or_default();

    let model = parsed.get("model").and_then(|v| match v {
        YamlValue::Scalar(s) => Some(s.clone()),
        _ => None,
    });

    let tools = parsed.get("tools").and_then(|v| match v {
        YamlValue::List(items) => Some(items.clone()),
        YamlValue::Scalar(s) => Some(vec![s.clone()]),
    });

    let max_budget_usd = parsed
        .get("max_budget_usd")
        .and_then(|v| match v {
            YamlValue::Scalar(s) => s.parse::<f64>().ok(),
            _ => None,
        })
        .filter(|v| v.is_finite());

    Ok(ParsedSkillDefinition {
        name,
        description,
        model,
        tools,
        max_budget_usd,
        instructions: body.to_string(),
        raw_content: content.to_string(),
    })
}

/// Load and parse a SKILL.md file from disk.
pub fn load_skill_file(file_path: &str) -> Option<ParsedSkillDefinition> {
    let path = Path::new(file_path);
    if !path.exists() {
        return None;
    }

    match std::fs::read_to_string(path) {
        Ok(content) => match parse_skill_md(&content) {
            Ok(parsed) => Some(parsed),
            Err(e) => {
                warn!(target: "skill-parser", "Failed to parse {}: {}", file_path, e);
                None
            }
        },
        Err(e) => {
            warn!(target: "skill-parser", "Failed to read {}: {}", file_path, e);
            None
        }
    }
}

/// Write a skill definition to a SKILL.md file.
pub fn write_skill_file(dir_path: &str, definition: &ParsedSkillDefinition) {
    let dir = Path::new(dir_path);
    if !dir.exists() {
        let _ = std::fs::create_dir_all(dir);
    }

    let mut frontmatter = format!("---\nname: {}\n", definition.name);
    frontmatter.push_str(&format!("description: >\n  {}\n", definition.description));

    if let Some(ref model) = definition.model {
        frontmatter.push_str(&format!("model: {}\n", model));
    }

    if let Some(ref tools) = definition.tools {
        if !tools.is_empty() {
            frontmatter.push_str("tools:\n");
            for tool in tools {
                frontmatter.push_str(&format!("  - {}\n", tool));
            }
        }
    }

    if let Some(budget) = definition.max_budget_usd {
        frontmatter.push_str(&format!("max_budget_usd: {}\n", budget));
    }

    frontmatter.push_str("---\n\n");

    let content = format!("{}{}", frontmatter, definition.instructions);
    let file_path = dir.join("SKILL.md");
    let _ = std::fs::write(file_path, content);
}

/// Scan a directory for SKILL.md files (one level deep).
///
/// Returns an array of scanned skills for each subdirectory containing
/// a valid SKILL.md file.
pub fn scan_skills_directory(base_path: &str) -> Vec<ScannedSkill> {
    let base = Path::new(base_path);
    if !base.exists() {
        return Vec::new();
    }

    let entries = match std::fs::read_dir(base) {
        Ok(entries) => entries,
        Err(_) => return Vec::new(),
    };

    let mut results = Vec::new();

    for entry in entries.flatten() {
        if !entry.file_type().map(|ft| ft.is_dir()).unwrap_or(false) {
            continue;
        }

        let dir_name = entry.file_name().to_string_lossy().to_string();
        let skill_file = entry.path().join("SKILL.md");
        let skill_file_str = skill_file.to_string_lossy().to_string();

        if let Some(parsed) = load_skill_file(&skill_file_str) {
            results.push(ScannedSkill {
                dir_name,
                path: entry.path().to_string_lossy().to_string(),
                parsed,
            });
        }
    }

    results
}

// ── Simple YAML parser (internal) ──────────────────────────

#[derive(Debug, Clone)]
enum YamlValue {
    Scalar(String),
    List(Vec<String>),
}

fn parse_simple_yaml(yaml: &str) -> std::collections::HashMap<String, YamlValue> {
    let mut result = std::collections::HashMap::new();
    let lines: Vec<&str> = yaml.lines().collect();
    let mut current_key = String::new();
    let mut current_value = String::new();
    let mut in_multiline = false;
    let mut in_list = false;
    let mut list_items: Vec<String> = Vec::new();

    for line in &lines {
        let trimmed = line.trim();

        if in_list {
            if let Some(item) = trimmed.strip_prefix("- ") {
                list_items.push(item.trim().to_string());
                continue;
            } else {
                // End of list
                result.insert(current_key.clone(), YamlValue::List(list_items.clone()));
                in_list = false;
                list_items.clear();
            }
        }

        if in_multiline {
            if trimmed.is_empty()
                || (!line.starts_with(' ')
                    && !line.starts_with('\t')
                    && trimmed.contains(':'))
            {
                // End of multiline
                result.insert(
                    current_key.clone(),
                    YamlValue::Scalar(current_value.trim().to_string()),
                );
                in_multiline = false;
            } else {
                current_value.push(' ');
                current_value.push_str(trimmed);
                continue;
            }
        }

        // Parse key: value
        if let Some(colon_idx) = trimmed.find(':') {
            if colon_idx > 0 {
                current_key = trimmed[..colon_idx].trim().to_string();
                let value = trimmed[colon_idx + 1..].trim().to_string();

                if value == ">" || value == "|" {
                    in_multiline = true;
                    current_value = String::new();
                } else if value.is_empty() {
                    // Could be start of a list
                    in_list = true;
                    list_items = Vec::new();
                } else {
                    result.insert(current_key.clone(), YamlValue::Scalar(value));
                }
            }
        }
    }

    // Handle trailing values
    if in_multiline && !current_key.is_empty() {
        result.insert(
            current_key.clone(),
            YamlValue::Scalar(current_value.trim().to_string()),
        );
    } else if in_list && !current_key.is_empty() {
        result.insert(current_key.clone(), YamlValue::List(list_items));
    }

    result
}
